File: app/src/main/java/com/example/eldhopjames/recyclercardviewsample/adapter/RecyclerAdapter.java
package com.example.eldhopjames.recyclercardviewsample.adapter;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.example.eldhopjames.recyclercardviewsample.R;
import com.example.eldhopjames.recyclercardviewsample.interfaces.OnItemClickListener;
import com.example.eldhopjames.recyclercardviewsample.modelClass.ModelClass;
import com.example.eldhopjames.recyclercardviewsample.viewHolders.EmptyViewHolder;
import com.example.eldhopjames.recyclercardviewsample.viewHolders.EvenViewHolder;
import com.example.eldhopjames.recyclercardviewsample.viewHolders.OddViewHolder;

import java.util.ArrayList;
import java.util.List;

/**
 * needs RecyclerAdapter and View Holder
 *      View Holder : used to caches the data and no need to write findViewById for each items
 *      RecyclerAdapter : populates the item  and data about it , data comes from n/w or from DB
 * We create an interface to pass the values of the item clicked into activity
 *
 */
public class RecyclerAdapter extends RecyclerView.Adapter<RecyclerView.ViewHolder> {


    private List<ModelClass> mListItems = new ArrayList<>();
    private Context mContext;
    private int totalListItems = 0;
    private OnItemClickListener mListener;

    //constructor
    public RecyclerAdapter( Context context) { // constructor
        this.mContext = context;
    }

    /**
     * interface will forward our click from adapter to our main activity
     */

    public void setOnItemClickListener(OnItemClickListener listener) {
        mListener = listener;
    }

    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) { // this method calls when ever our view method is created , ie; the instance of ViewHolder class is created
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View view;
        switch (viewType) {
            case 0:
                view = inflater
                        .inflate(R.layout.even_list_item, parent, false);
                return new OddViewHolder(view, mListener);
            case 1:
                view = inflater
                        .inflate(R.layout.odd_list_item, parent, false);
                return new EvenViewHolder(view, mListener);
            default:
                view = inflater
                        .inflate(R.layout.item_empty, parent, false);
                return new EmptyViewHolder(view);


//                //NOTE : use this if items height and width not following the match_parent or wrap_content
//                RecyclerView.LayoutParams layoutParams = new RecyclerView.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT,ViewGroup.LayoutParams.WRAP_CONTENT);
//                view.setLayoutParams(layoutParams);
        }
    }

    @Override
    public int getItemViewType(int position) {
        if (mListItems.get(position).getType() == 0){
            return 0;
        } else if (mListItems.get(position).getType() == 1) {
            return 1;
        }
        return super.getItemViewType(position);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position) { //populate the data into the list_item (View Holder), as we scroll
        EvenViewHolder evenViewHolder;
        OddViewHolder oddViewHolder;

        if (holder instanceof EvenViewHolder){ // checks which viewHolder is it
            evenViewHolder = (EvenViewHolder) holder;
            setEvenViewHolder (evenViewHolder,position);
        }
        if (holder instanceof OddViewHolder){
            oddViewHolder = (OddViewHolder) holder;
            setOddViewHolder(oddViewHolder,position);
        }
    }

    @Override
    public int getItemCount() {
        return mListItems == null ? 0 : mListItems.size();
    }

    private void setOddViewHolder(OddViewHolder oddViewHolder, int position) {
        ModelClass listItem = mListItems.get(position);
        if (listItem != null) {
            oddViewHolder.headTv.setText(listItem.getHead());
            oddViewHolder.descTv.setText(listItem.getDesc());
        }
    }

    private void setEvenViewHolder(EvenViewHolder evenViewHolder, int position) {
        ModelClass listItem = mListItems.get(position);
        if (listItem != null) {
            evenViewHolder.headTv.setText(listItem.getHead());
            evenViewHolder.descTv.setText(listItem.getDesc());
        }
    }

    //-------------------------------------Manipulating RecyclerView--------------------------------//
    public void clearData() {
        if (!mListItems.isEmpty()) {
            mListItems.clear();
            notifyDataSetChanged();
            totalListItems = 0;
        }
    }

    public void addItemRange(List<ModelClass> items) {
        if (items != null) {
            int position = totalListItems;
            mListItems.addAll(position, items);
            notifyItemRangeInserted(position, items.size()); //items.size()+1 to disable auto scroll
            totalListItems = mListItems.size();
        }
    }

    public void addItemRangeInPosition(List<ModelClass> items, int position) {
        if (items != null) {
            if (totalListItems >= position) {
                mListItems.addAll(position, items);
                notifyItemRangeInserted(position, items.size());
            } else {
                mListItems.addAll(items);
                notifyItemRangeInserted(totalListItems - 1, items.size());
            }
            totalListItems = mListItems.size();
        }
    }

    public void addItem(ModelClass item, int position) {
        if (item != null) {
            if (totalListItems >= position) {
                mListItems.add(position, item);
                notifyItemInserted(position);
            } else {
                mListItems.add(item);
                notifyItemInserted(totalListItems - 1);
            }
            ++totalListItems;
        }
    }

    public void removeItem(int position) {
        if (totalListItems >= position) {
            mListItems.remove(position);
            notifyItemRemoved(position);
        } else {
            int newPosition = totalListItems - 1;
            mListItems.remove(newPosition);
            notifyItemRemoved(newPosition);
        }
        --totalListItems;
    }
}

File: app/src/main/java/com/example/eldhopjames/recyclercardviewsample/adapter/RecyclerAdapter.kt
package com.example.eldhopjames.recyclercardviewsample.adapter

import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import com.example.eldhopjames.recyclercardviewsample.databinding.EvenListItemBinding
import com.example.eldhopjames.recyclercardviewsample.databinding.ItemEmptyBinding
import com.example.eldhopjames.recyclercardviewsample.databinding.OddListItemBinding
import com.example.eldhopjames.recyclercardviewsample.modelClass.ModelClass
import com.example.eldhopjames.recyclercardviewsample.viewHolders.EmptyViewHolder
import com.example.eldhopjames.recyclercardviewsample.viewHolders.EvenViewHolder
import com.example.eldhopjames.recyclercardviewsample.viewHolders.OddViewHolder

/**
 * needs RecyclerAdapter and View Holder
 * View Holder : used to caches the data and no need to write findViewById for each items
 * RecyclerAdapter : populates the item  and data about it , data comes from n/w or from DB
 * We create an interface to pass the values of the item clicked into activity
 *
 */
class RecyclerAdapter :
    ListAdapter<ModelClass, ViewHolder>(DiffCallback()) {
    private var listener: ((ModelClass, Int) -> Unit)? = null

    /**
     * lambda will forward our click from adapter to our activity/fragment
     * NOTE : we can pass lambda via constructor of the adapter also
     */
    fun setOnContentClickListener(listener: (ModelClass, Int) -> Unit) {
        this.listener = listener
    }

    class DiffCallback : DiffUtil.ItemCallback<ModelClass>() {
        override fun areItemsTheSame(oldItem: ModelClass, newItem: ModelClass): Boolean =
            oldItem.primaryKey == newItem.primaryKey

        override fun areContentsTheSame(oldItem: ModelClass, newItem: ModelClass): Boolean =
            oldItem == newItem

    }

    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder { // this method calls when ever our view method is created , ie; the instance of ViewHolder class is created
        val inflater = LayoutInflater.from(parent.context)
        val binding: Any
        return when (viewType) {
            VIEW_TYPE_ODD -> {
                binding = EvenListItemBinding.inflate(
                    inflater,
                    parent, false
                )
                EvenViewHolder(binding, listener)
            }
            VIEW_TYPE_EVEN -> {
                binding = OddListItemBinding.inflate(
                    inflater,
                    parent, false
                )
                OddViewHolder(binding, listener)
            }
            else -> {
                binding = ItemEmptyBinding.inflate(
                    inflater,
                    parent, false
                )
                EmptyViewHolder(binding)
            }
        }
    }

    //TODO: Migrate in favor of merge adapter
    override fun getItemViewType(position: Int): Int {
        if (getItem(position).type == 0) {
            return VIEW_TYPE_ODD
        } else if (getItem(position).type == 1) {
            return VIEW_TYPE_EVEN
        }
        return super.getItemViewType(position)
    }

    override fun onBindViewHolder(
        holder: ViewHolder,
        position: Int
    ) { //populate the data into the list_item (View Holder), as we scroll
        when (holder) {
            is EvenViewHolder -> holder.bindData(getItem(position))
            is OddViewHolder -> holder.bindData(getItem(position))
        }
    }


    //-------------------------------------Manipulating RecyclerView--------------------------------//

    /**
     * Submitting a arrayList into a particular position (if position is not valid, we will append at the end) by default appending at the end
     * */
    fun submitListItem(newDataList: MutableList<ModelClass>?, position: Int = itemCount) {
        if (!newDataList.isNullOrEmpty()) {
            val newItems: MutableList<ModelClass> = currentList.toMutableList()
            if (itemCount >= position) {
                newItems.addAll(position, newDataList)
            } else {
                newItems.addAll(newDataList)
            }
            submitList(newItems)
        }
    }

    /**
     * clearing existing list and submit fresh list
     * */
    fun replaceData(newDataList: MutableList<ModelClass>?) {
        if (!newDataList.isNullOrEmpty()) {
            submitList(newDataList)
        }
    }

    /**
     * Submitting a object into a particular position (if position is not valid, we will append at the end)or appending at the end
     * */
    fun addItem(newData: ModelClass?, position: Int = itemCount) {
        if (newData != null) {
            val newItems: MutableList<ModelClass> = currentList.toMutableList()
            if (itemCount > position) {
                newItems.add(position, newData)
            } else {
                newItems.add(newData)
            }
            submitList(newItems)
        }
    }

    /**
     * clearing of an object from a particular position (if position is not valid, we will remove from the end) or from last adapter
     * */
    fun removeItem(position: Int = itemCount - 1) {
        val newItems: MutableList<ModelClass> = currentList.toMutableList()
        if (itemCount > position) {
            newItems.removeAt(position)
        } else {
            newItems.removeAt(itemCount - 1)
        }
        submitList(newItems)
    }

    companion object {
        private const val VIEW_TYPE_ODD = 1
        private const val VIEW_TYPE_EVEN = 2
    }
}
